Fix: Responsive settings for Section Templates wont fully save [ED-11103]

maork-elementor opened the pull request.

maork-elementor committed:
Fix: Responsive settings for Section Templates won't fully save when additional breakpoints are enabled [ED-11103]

maork-elementor committed:
Fix: Responsive settings for Section Templates wont fully save when additional breakpoints are enabled [ED-11103]

maork-elementor committed:
Compatibility support to existing ga pro version

maork-elementor committed:
Refactor essential CSS files.

- Refactor CSS code to improve readability and maintainability
- Consolidate styles from multiple files into base.php

maork-elementor committed:
Merge remote-tracking branch 'elementor/main' into ED-11103_02

maork-elementor committed:
Refactor Base CSS for better performance.

- Update base.php in core/files/css
- Improve performance and fix bugs

maork-elementor committed:
Remove comment

maork-elementor committed:
tmp

maork-elementor committed:
Merge remote-tracking branch 'elementor/main' into ED-11103

maork-elementor committed:
test

maork-elementor committed:
test

maork-elementor committed:
update

maork-elementor committed:
Merge remote-tracking branch 'origin/ED-11103' into ED-11103_02

maork-elementor committed:
update

maork-elementor committed:
update comment

maork-elementor committed:
update

maork-elementor committed:
test

maork-elementor committed:
update

maork-elementor committed:
update

maork-elementor committed:
update

maork-elementor committed:
fix

maork-elementor committed:
lint

maork-elementor committed:
fix

maork-elementor committed:
lint

maork-elementor committed:
test: Update responsive controls tests

- Update file path for template in `template` object

matipojo commented on core/files/css/base.php:
```diff
@@ -643,6 +643,34 @@ protected function is_update_required() {
+	 */
+	private function clear_stacks_on_additional_breakpoints_active() {
+		$is_edit_mode = Plugin::$instance->editor->is_edit_mode();
+		if ( ! $is_edit_mode && method_exists( Plugin::$instance->controls_manager, 'clear_stack_cache' ) ) {
+			$currnet_post_id = null;
+			if ( method_exists( $this, 'get_post_id' ) ) {
```
Why do you need to check if `method_exists`?

maork-elementor replied:
Because not in every iteration, we actually have `post_id`

matipojo replied:
A parent class shouldn't know about his children, so this code should be moved to the child that always has the method

maork-elementor replied:
Fixed.

matipojo commented on includes/managers/controls.php:
```diff
@@ -301,8 +301,11 @@ class Controls_Manager {
 	 *
 	 * @var array
 	 */
+	// TODO: Remove in the future. - When Pro will remove the template class use in location manager.
```
It's a private property, and it's safe to remove it

maork-elementor replied:
In pro version, there is a use in :

https://github.com/elementor/elementor-pro/blob/c35e2084b5884c3b92de9da3817655a0c10e2432/modules/theme-builder/classes/locations-manager.php

of 

https://github.com/elementor/elementor-pro/blob/6ab8d9dc94328c75ba80396c64992fe51dd63137/modules/theme-builder/files/css/template.php

with the next function 

https://github.com/elementor/elementor/blob/3994d6d3c801e2234f60a0f150630061fd54f20c/includes/managers/controls.php#L894

That has support only for one time flag value update per page render per document.
So I need to support both ways until we will remove the support from the Pro version.
Then the location manager in pro will inherit the new logic from base that support multipal page ID's support.

https://github.com/maork-elementor/elementor/blob/b36f98759cad724e413d10200400ae1bb2aabd89/includes/managers/controls.php#L906

matipojo replied:
It can't be an external direct use of a private property.

maork-elementor replied:
In my understanding, we will continue to support each other in both directions?
I have removed the comment.

matipojo commented on core/files/css/base.php:
```diff
@@ -643,6 +643,34 @@ protected function is_update_required() {
+	 */
+	private function clear_stacks_on_additional_breakpoints_active() {
+		$is_edit_mode = Plugin::$instance->editor->is_edit_mode();
+		if ( ! $is_edit_mode && method_exists( Plugin::$instance->controls_manager, 'clear_stack_cache' ) ) {
```
Why do you need to check if `method_exists`?

maork-elementor replied:
Removed.

Commit:
Merge branch 'main' into ED-11103_02

maork-elementor committed:
Merge remote-tracking branch 'elementor/main' into ED-11103_02

maork-elementor committed:
Refactor base.php file in core directory

maork-elementor committed:
Refactor base.php file in core directory

maork-elementor committed:
Merge remote-tracking branch 'origin/ED-11103_02' into ED-11103_02

maork-elementor committed:
chroe: Improve lazy loading functionality in `lazyload` module.

- Remove method exist on `get_post_id`

maork-elementor committed:
test

maork-elementor committed:
add test

maork-elementor committed:
Merge branch 'main' into ED-11103_02

maork-elementor committed:
test

maork-elementor committed:
test

maork-elementor committed:
Merge remote-tracking branch 'origin/ED-11103_02' into ED-11103_02

maork-elementor committed:
slow ci

maork-elementor committed:
test

maork-elementor committed:
test

maork-elementor committed:
test

maork-elementor committed:
Merge branch 'main' into ED-11103_02

maork-elementor committed:
add shortcode test

maork-elementor committed:
ci fix

maork-elementor committed:
test

maork-elementor committed:
ci not working

maork-elementor committed:
Merge branch 'main' into ED-11103_02

maork-elementor committed:
test

maork-elementor committed:
test

maork-elementor committed:
Merge remote-tracking branch 'origin/ED-11103_02' into ED-11103_02

maork-elementor committed:
remove tests

maork-elementor committed:
fix

maork-elementor committed:
Merge branch 'main' into ED-11103_02

maork-elementor committed:
Merge branch 'main' into ED-11103_02

maork-elementor commented on core/files/css/base.php:
```diff
@@ -643,6 +643,42 @@ protected function is_update_required() {
 		return false;
 	}
 
+	/**
```
Nice, looks great.

maork-elementor committed:
refactor: Refactor core files and includes managers

- Refactor CSS in [core/files/css/base.php] for improved readability and maintenance
- Update code comments in [core/files/css/base.php] for better code documentation and clarity

maork-elementor committed:
Merge branch 'main' into ED-11103_02

maork-elementor committed:
Revert changes.

maork-elementor committed:
Merge remote-tracking branch 'origin/ED-11103_02' into ED-11103_02

matipojo commented on core/files/css/base.php:
```diff
@@ -643,6 +643,28 @@ protected function is_update_required() {
+	 *
+	 * @access private
+	 */
+	private function clear_stacks() {
```
I think the full logic is not related to here, so please move the logic to the app/module layers.

You can work with hooks, for example, on `parse_content`, trigger an action: 
```
do_action( 'elementor/css_file/parse_content' );
```
Then from `additional_custom_breakpoints` experiment, run logic and then `clear_stack_cache`:
```
add_action( 'elementor/css_file/parse_content', function() { 
	if...
	Plugin::$instance->controls_manager->clear_stack_cache( $handle_id );
});
```

maork-elementor committed:
Merge remote-tracking branch 'elementor/main' into ED-11103_02

maork-elementor committed:
Requested changes

matipojo commented on core/files/css/base.php:
```diff
@@ -652,6 +652,9 @@ protected function is_update_required() {
 	 */
 	protected function parse_content() {
+
+		do_action( 'elementor/css_file/parse_content' );
```
```suggestion
		do_action( 'elementor/css_file/parse_content', $this );
```

maork-elementor committed:
Merge remote-tracking branch 'elementor/main' into ED-11103_02

matipojo commented on core/breakpoints/manager.php:
```diff
@@ -534,4 +534,36 @@ private static function get_stylesheet_templates() {
+	}
+
+	public function __construct() {
+		add_action( 'elementor/css_file/parse_content', function() {
+			$handle_id = $this->get_file_handle_id();
```
```suggestion
		add_action( 'elementor/css_file/parse_content', function( $css_file ) {
			$handle_id = $css_file->get_file_handle_id();
```

maork-elementor replied:
As a result of these changes, it looks like this :
Template widget and shortcode widget stop generating responsive controls, like in this initial state. 

I'm looking into this at the moment

matipojo commented on core/breakpoints/manager.php:
```diff
@@ -534,4 +534,36 @@ private static function get_stylesheet_templates() {
 
 		return apply_filters( $replacement_hook, $templates );
 	}
+
+	/**
+	 * Get file handle ID.
+	 *
+	 * Retrieve the handle ID for the post CSS file.
+	 *
+	 * @since 3.15.0
+	 * @access protected
+	 *
+	 * @return string CSS file handle ID.
+	 */
+	protected function get_file_handle_id() {
+		$current_document = Plugin::$instance->documents->get_current();
+		if ( $current_document ) {
+			$post_id = $current_document->get_main_id();
+			return 'elementor-post-' . $post_id;
+		}
+
+		return false;
+	}
+
```
Remove

maork-elementor committed:
Update core/files/css/base.php

Co-authored-by: Mati Horowitz <21468434+matipojo@users.noreply.github.com>

maork-elementor committed:
Update core/breakpoints/manager.php

Co-authored-by: Mati Horowitz <21468434+matipojo@users.noreply.github.com>

matipojo commented on core/files/css/base.php:
```diff
@@ -652,6 +652,9 @@ protected function is_update_required() {
 	 * @access protected
 	 */
 	protected function parse_content() {
+
```
```suggestion
```

matipojo commented on core/breakpoints/manager.php:
```diff
@@ -534,4 +534,36 @@ private static function get_stylesheet_templates() {
+			if ( ! $is_edit_mode && $handle_id && $additional_breakpoints_active ) {
+				Plugin::$instance->controls_manager->clear_stacks( $handle_id );
+			}
+		});
```
PHPCS

maork-elementor replied:
PHPCS saying everything is linted.

maork-elementor committed:
Remove

maork-elementor committed:
lint

maork-elementor committed:
update

maork-elementor committed:
Update

maork-elementor committed:
Add tests

maork-elementor committed:
test for default

maork-elementor committed:
Merge branch 'main' into ED-11103_02

matipojo commented on includes/managers/controls.php:
```diff
@@ -881,8 +881,12 @@ public function remove_control_from_stack( $stack_id, $control_id ) {
-	public function has_stacks_cache_been_cleared() {
-		return $this->has_stacks_cache_been_cleared;
+	public function has_stacks_cache_been_cleared( $handle_id = 'default' ) {
+		if ( ! isset( $this->has_post_stacks_cache_been_cleared[ $handle_id ] ) ) {
+			return false;
+		}
+
+		return true;
```
```suggestion
		return isset( $this->has_post_stacks_cache_been_cleared[ $handle_id ] );
```

matipojo commented on includes/managers/controls.php:
```diff
@@ -1174,4 +1179,24 @@ public function add_custom_attributes_controls( Controls_Stack $controls_stack )
+	 */
+	public function clear_stacks( $handle_id ) {
+		if ( $handle_id ) {
+			$stack_cache_has_been_cleared = $this->has_stacks_cache_been_cleared( $handle_id );
+			if ( ! $stack_cache_has_been_cleared ) {
+				$this->clear_stack_cache( $handle_id );
+			}
```
```suggestion
			if ( ! $this->has_stacks_cache_been_cleared( $handle_id ) ) {
				$this->clear_stack_cache( $handle_id );
			}
```

matipojo commented on tests/phpunit/elementor/includes/managers/test-controls.php:
```diff
@@ -105,6 +105,72 @@ public function test_clear_stack_cache() {
+		$this->assertFalse( $stack_cache_has_been_cleared );
+
+		// Act
+		if ( ! $stack_cache_has_been_cleared ) {
+			Plugin::$instance->controls_manager->clear_stack_cache( $handle_id );
+		}
```
It's already asserted as false.
```suggestion
		Plugin::$instance->controls_manager->clear_stack_cache( $handle_id );
```

maork-elementor committed:
Update includes/managers/controls.php

Co-authored-by: Mati Horowitz <21468434+matipojo@users.noreply.github.com>

maork-elementor committed:
Update includes/managers/controls.php

Co-authored-by: Mati Horowitz <21468434+matipojo@users.noreply.github.com>

maork-elementor committed:
Update tests/phpunit/elementor/includes/managers/test-controls.php

Co-authored-by: Mati Horowitz <21468434+matipojo@users.noreply.github.com>
